### ISSUE TITLE
Data for Mexico MX

### ISSUE THREAD

**Issue by alixunderplatz** [electricitymap/electricitymap-contrib#1242]
Promising news!

There is total generation (no breakdown yet), load data and power flow data for **Mexico**! 
Seems like [this site](https://www.gob.mx/cenace/articulos/graficas-de-demanda-y?idiom=es) has been online since 15th March, so that's an accidental find :)

Power flows and total generation and consumption per region (hit button on the right to switch):
http://www.cenace.gob.mx/Paginas/Publicas/Info/DemandaRegional.aspx

![image](https://user-images.githubusercontent.com/25743609/37621631-c3d10f04-2bbf-11e8-9c99-24449a74084f.png)

And generation and consumption as a time series (seems like it is updated every 15 min, see timestamp on the left):
http://www.cenace.gob.mx/GraficaDemanda.aspx

![image](https://user-images.githubusercontent.com/25743609/37623487-2fc0fae4-2bc5-11e8-9bcc-612749002b9d.png)

All live exchanges with the U.S. in the North are there, also with Belize (new) and Guatemala in the south.

Seems like the two regions on the left part of another system/synchronous area: "National Interconnected System, does not include the regions of Baja California and Baja California Sur."
So California is only linked with these and not with the Mexican "mainland", but Texas is.

_ _ _ _

Seems like this is the open data platform of the Mexican government:
https://datos.gob.mx/libreusomx

Maybe there will be or is more info on breakdown, so please take a look!

Terms of use: https://datos.gob.mx/libreusomx

**Comment by systemcatch** [comment]
Going on http://www.cenace.gob.mx/Paginas/Publicas/Info/DemandaRegional.aspx and clicking 'intercambio entre regiones' (exchange between regions) seems to show live flows between Mexico and the USA. We could use this for #959!

**Comment by alixunderplatz** [comment]
@systemcatch right, works for CAISO imports! 
The MX region "Baja California" in the upper left corner is not connected with the rest of the Mexikan mainland, so the influence on CAISO will be really low (24 MW at the moment, I've found up to ~100 MW on EIA's site in March) 

![image](https://user-images.githubusercontent.com/25743609/37722182-a408f098-2d2b-11e8-976f-5d82eed4d9d8.png)

**Comment by brunolajoie** [comment]
What about adding it to the map as 100% "unknown", taking 2017 average mix to compute CO2 intensity of "unknown", and iterate from there when we'll have the proper breakdown per fuel?

**Comment by alixunderplatz** [comment]
@brunolajoie do you mean entire Mexico or just the upper left corner?

**Comment by brunolajoie** [comment]
I meant all regions 😃 (and the exchanges btw them).
I know it hasn't been the usual practice, but Mexico is a major country, and I'm wondering if adding it to the map with an unknown power mix might not help convince them to disclose the breakdown as a next step! thinking out loud here!

**Comment by alixunderplatz** [comment]
According to [this source](https://www.pv-magazine.com/2018/01/09/mexicos-pv-capacity-reached-460-mw-at-the-end-of-june/), there were around 4 GW of wind and 0.46 GW of PV by middle of 2017 in MX.
Installed PV could have doubled in the meantime and wind can contribute up to 1/7 of the total demand, which will "break" the intensity factor. Their hydro generation will most likely follow the load scheme as well.

Of course the "unknown" assumption will make some areas with "total generation only"-data look horribly bad, which could raise the desire to publish the actual generation mix. But it will also negatively influence the intensity of zones in importing from MX.

_ _ _ _

In general, I support including generation data with an unknown intensity, because it is a way to make people become aware of when consumption in their area is the highest (which often is the equivalent to "the dirtiest" or "most stress for the system"="higher risk of outage"). But this is rather an option to further develop the functions of electricitymap, aside from the main goal of showing carbon intensity.
Coloring them in stripes on the map to highlight the "unknown" live generation mix could be a way to do it. _(Just leaving some other candiadtes here: Japan, South Korea, Iceland (demand = generation in these 3), The Netherlands)_

**Comment by systemcatch** [comment]
I'm looking at adding the exchange between California and Baja California Norte, but it seems like the time is stuck at 12:00:00 while the exchange value is changing. Does anyone else see this?

**Comment by alixunderplatz** [comment]
FYI: I accidentally found a website with some more general and regional info on the profile of mexican electricity generation and the location of power plants - images are from around 2016:
http://www.conacytprensa.mx/index.php/tecnologia/energia/6597-cenace-energia-potencia-electrica

![image](https://user-images.githubusercontent.com/25743609/43990889-4dc5896a-9d62-11e8-8a5a-e12d24e7e812.png)

![image](https://user-images.githubusercontent.com/25743609/43990885-4139faa0-9d62-11e8-9a17-2d82c0fac59b.png)

**Comment by systemcatch** [comment]
MX exchanges are failing due to SSL verification of of the sites certificate.

**Comment by alixunderplatz** [comment]
Total generation for all Mexican zones could still be added to the map as well (like for Japan), although there is no value added regarding carbon intensity. But it's always interesting just to have it and visualize it.

https://www.cenace.gob.mx/Paginas/Publicas/Info/DemandaRegional.aspx

![image](https://user-images.githubusercontent.com/25743609/45777929-71124700-bc57-11e8-8923-68a7502efa32.png)

**Comment by cuauhtemoc11** [comment]
There hasn't been any progress for this country in almost a year. Maybe joining the regions would make it easier for the time being?

**Comment by systemcatch** [comment]
@cuauhtemoc11 I don't see how that makes it easier, we lack any production data for Mexico.

**Comment by scriptator** [comment]
I found a page which looks to me like it has production data for the whole country: https://www.cenace.gob.mx/SIM/VISTA/REPORTES/EnergiaGenLiqAgregada.aspx

What do you think - is this valuable?

**Comment by jarek** [comment]
That looks quite good, however do I see correctly that it's currently (mid-October) only data for September available?

**Comment by scriptator** [comment]
Yes, sadly you are right. I did not notice that before because I had only looked at the publication date. Would it be still interesting to have a parser for historical data?

**Comment by alixunderplatz** [comment]
@scriptator nice! imo, it's never wrong to have national historical data available.
Maybe @corradio could include this data in the visual review for the entire year 2019 ;)
(Like for 2018: https://www.youtube.com/watch?v=GxB91InvxiY)

**Comment by scriptator** [comment]
Cool, then I am going to give this a shot. Nice Video btw - I didn't know that so far.

**Comment by scriptator** [comment]
In the fetch_* functions of a parser, how much data before or after the provided target_datetime should or can be returned? As fetching the CSV file takes quite a bit of time and returns a whole month it would make sense here to return as much as possible IMO.

**Comment by corradio** [comment]
For historical data it doesn't really matter (the backend will adapt)
For real-time data the parser should be fast enough so 24hrs in practice is reasonable.

**Comment by martin-laurent** [comment]
There seems to be a new website by the government with a few CSV files to download: https://www.cenace.gob.mx/APSIM.aspx
I didn't got the time to look in which data exactly is relevant (don't speak Spanish)

**Comment by madsnedergaard** [comment]
Thanks for all the valuable information shared here! 🙌

Copying data from #3032 in here so it's all in one place:

> I found the actual demand here which is 24 hours delayed: [cenace.gob.mx/Paginas/SIM/Reportes/EstimacionDemandaReal.aspx](https://www.cenace.gob.mx/Paginas/SIM/Reportes/EstimacionDemandaReal.aspx)
>
> And an hourly breakdown which is updated monthly:
[cenace.gob.mx/Paginas/SIM/Reportes/EnergiaGeneradaTipoTec.aspx](https://www.cenace.gob.mx/Paginas/SIM/Reportes/EnergiaGeneradaTipoTec.aspx)
>
> There's also an estimated forecasts, but now sure that we really want to rely on that:
[cenace.gob.mx/Paginas/SIM/Reportes/PronosticosDemanda.aspx](https://www.cenace.gob.mx/Paginas/SIM/Reportes/PronosticosDemanda.aspx)


We're closing this issue as it is not currently possible to build a real-time parser given the data provided here.
If you find new information that might help us get it done, please comment and reopen the issue [when a parser is buildable](https://github.com/electricityMap/electricitymap-contrib/wiki/Technical-requirements-for-parser-data) :)

**Comment by ToasterDEV** [comment]
Hey there!

Just found out not long ago about this project, and I think I might have some good news!

Going through the previously mentioned website I was able to find a new [URL](https://www.cenace.gob.mx/GraficaDemanda.aspx) that just so happens to have an AJAX endpoint that outputs clean JSON data for hourly reports [here](https://www.cenace.gob.mx/GraficaDemanda.aspx/obtieneValoresTotal).

According to the filtering available on the script that generates the table, changing the parameter "gerencia" even outputs different regions, and I was able to successfully extract data using the following cURL command:

`curl -X POST "https://www.cenace.gob.mx/GraficaDemanda.aspx/obtieneValoresTotal" -H "Content-Type: application/json; charset=utf-8" -d '{"gerencia":"10"}'`

Where "gerencia" is 10 as the default (national) value.

Would this be enough to integrate into the service?
